api/v1/verticadb_webhook.go

```diff
@@ -1701,6 +1728,199 @@ func (v *VerticaDB) checkSandboxPrimary(allErrs field.ErrorList, oldObj *Vertica
+		if newSandbox.Shutdown {
+			for _, subclusterName := range oldSandbox.Subclusters {
+				oldSubcluster := oldSubclusterMap[subclusterName.Name]
+				if vmeta.GetShutdownDrivenBySandbox(oldSubcluster.Annotations) {
+					newSubcluster, oldSclusterPersist := newSubclusterMap[subclusterName.Name]
+					if oldSclusterPersist && oldSubcluster.Shutdown != newSubcluster.Shutdown {
+						index := newSubclusterIndexMap[subclusterName.Name]
+						p := field.NewPath("spec").Child("subclusters")
+						err := field.Invalid(p.Index(index).Child("shutdown"),
+							newSubcluster.Shutdown,
+							fmt.Sprintf("cannot change Shutdown field for subcluster %q whose shutdown is driven by sandbox %q",
+								subclusterName.Name, oldSandbox.Name))
+						allErrs = append(allErrs, err)
+					}
+				}
+			}
+		}
+	}
+	return allErrs
+}
+
+// ensures a sandbox's image is not updated after the sandbox or any of its subclusters are shut down
+func (v *VerticaDB) validateShutdownSandboxImage(old runtime.Object, allErrs field.ErrorList) field.ErrorList {
+	oldObj := old.(*VerticaDB)
+	persistSandboxes := v.findPersistSandboxes(oldObj)
+	oldSandboxMap := oldObj.GenSandboxMap()
+	newSandboxMap := v.GenSandboxMap()
+	newSandboxIndexMap := v.GenSandboxIndexMap()
+	newSubclusterMap := v.GenSubclusterMap()
+	statusSClusterIndexMap := v.GenStatusSClusterIndexMap()
+	for _, sandboxName := range persistSandboxes {
+		oldSandbox := oldSandboxMap[sandboxName]
+		newSandbox := newSandboxMap[sandboxName]
+		if oldSandbox.Image != newSandbox.Image {
+			newSandboxIndex := newSandboxIndexMap[sandboxName]
+			errMsgs := v.checkSboxForShutdown(newSandbox, newSubclusterMap, statusSClusterIndexMap)
+			if len(errMsgs) != 0 {
+				p := field.NewPath("spec").Child("sandboxes").Index(newSandboxIndex)
+				err := field.Invalid(p.Child("image"),
+					newSandbox.Image,
+					fmt.Sprintf("cannot change the image for sandbox %q because %q",
+						sandboxName, strings.Join(errMsgs, ",")))
+				allErrs = append(allErrs, err)
+			}
+		}
+	}
+	return allErrs
+}
+
+// ensure the sandbox to terminate is not shut down
+func (v *VerticaDB) validateTerminatingSandboxes(old runtime.Object, allErrs field.ErrorList) field.ErrorList {
+	oldObj := old.(*VerticaDB)
+	oldSandboxMap := oldObj.GenSandboxMap()
+	newSandboxMap := v.GenSandboxMap()
+	sandboxesToBeRemoved := vutil.MapKeyDiff(oldSandboxMap, newSandboxMap)
+	newSubclusterMap := v.GenSubclusterMap()
+	for _, sandboxToBeRemoved := range sandboxesToBeRemoved {
+		numOfSclustersInSbox := len(oldSandboxMap[sandboxToBeRemoved].Subclusters)
+		sclustersToTerminate := numOfSclustersInSbox
+		for _, subcluster := range oldSandboxMap[sandboxToBeRemoved].Subclusters {
+			if _, ok := newSubclusterMap[subcluster.Name]; ok {
+				sclustersToTerminate--
+			}
+		}
+		// sandboxToBeRemoved and all its subclusters are not found in new spec
+		if sclustersToTerminate == numOfSclustersInSbox {
+			if oldSandboxMap[sandboxToBeRemoved].Shutdown {
```

Comment:
This check should be at the beginning of the loop so you can skip to the next sandbox if the current sandbox has shutdown true.

Reply:
This if statement compares the number of subclusters to be removed with the number of subclusters in the sandbox.
If they equal, that means the sandbox will be terminated.
There is another for loop on the top of this if statement. It is used to calculate "**sclustersToTerminate**" used in the if statement.

Reply:
Again, if `oldSandboxMap[sandboxToBeRemoved].Shutdown` is false then there is no point in running all this code. Check `oldSandboxMap[sandboxToBeRemoved].Shutdown` at the beginning of the loop and if false skip and go to next sandbox. Something like:
```
for _, sandboxToBeRemoved := range sandboxesToBeRemoved {
       if !oldSandboxMap[sandboxToBeRemoved].Shutdown {
              continue
       }
```

Reply:
That is a good point. I will check oldSandboxMap[sandboxToBeRemoved].Shutdown earlier. If it is false, let it skip the rest of the current loop.